{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## imported required packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.metrics import roc_auc_score, auc, roc_curve\n",
    "from sklearn.model_selection import cross_val_score, train_test_split, StratifiedKFold, GridSearchCV\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import scipy.stats as stats\n",
    "from test_function import *\n",
    "\n",
    "f = open(\"/storage/home/jkl5991/group/dbnsfp/tabix_try/tabixindex.txt\")\n",
    "header = f.readline().strip().split('\\t')\n",
    "f.close()\n",
    "\n",
    "omit_model = pd.read_csv(\"/storage/home/jkl5991/work/project/not_conflict/all_model_new.tsv\", sep = '\\t')\n",
    "\n",
    "omit = pd.read_csv('unannotated_omit_std.tsv', sep = '\\t')\n",
    "dominant = pd.read_csv('dominant_std.tsv', sep = '\\t')\n",
    "recessive = pd.read_csv(\"recessive_std.tsv\",sep = '\\t')\n",
    "\n",
    "# the features we used in our original study\n",
    "original_column = ['SIFT_pred','LRT_pred', 'MA_pred', 'PROVEN_pred', 'SLR_score', 'SIFT_score','LRT_omega', \n",
    "                'MA_score', 'PROVEN_score', 'Grantham', 'HMMEntropy','HMMRelEntropy', 'PredRSAB', 'PredRSAI', \n",
    "                'PredRSAE','PredBFactorF', 'PredBFactorM', 'PredBFactorS', 'PredStabilityH','PredStabilityM', \n",
    "                'PredStabilityL', 'PredSSE', 'PredSSH','PredSSC', 'dscore', 'phyloP_pri', 'phyloP_mam','phyloP_ver',\n",
    "                   'RNA_seq','UNEECON']\n",
    "\n",
    "# the features for comparison (to other supervised score)\n",
    "features = [ 'MetaSVM_score', 'MetaLR_score', 'M-CAP_score', 'REVEL_score','ClinPred_score','CADD_phred_hg19',\n",
    "            'DANN_score','MPC','primateAI','FATHMM_score', 'VEST4_score', 'LIST-S2_score']\n",
    "\n",
    "\n",
    "\n",
    "allcolumn = original_column + features\n",
    "\n",
    "\n",
    "\n",
    "# a function that create a \"location\" column for future combination\n",
    "def createloc(df):\n",
    "    df['location'] = df['chr'] + '-' + (df['pos']-1).map(str)+ '-' + df['pos'].map(str)+ '-' + df['ref'].map(str)+ '-' + df['alt'].map(str)\n",
    "    df = df.drop(columns = ['chr', 'pos', 'ref', 'alt'])\n",
    "    return(df)\n",
    "\n",
    "\n",
    "def merge_with_original_df(originaldf, newdf):\n",
    "    newdf = createloc(newdf)\n",
    "    result = pd.merge(originaldf, newdf, on=['location'], how = 'inner')\n",
    "    return(result)\n",
    "\n",
    "### build de novo dataframe\n",
    "\n",
    "def find_denovo_pattern_new(readfile, find_pattern):\n",
    "    denovo_column = ['SampleID', 'StudyName', 'PubmedID', 'NumProbands', 'NumControls', 'SequenceType', 'PrimaryPhenotype', \n",
    "                 'Validation', 'Chr', 'Position', 'Variant', 'rsID', 'DbsnpBuild' , 'AncestralAllele', '1000GenomeCount', \n",
    "                 'ExacFreq', 'EspAaFreq', 'EspEaFreq', 'Transcript', 'condingDnaSize', 'Gene', 'FunctionClass', 'cDnaVariant', 'ProteinVariant', \n",
    "                 'Exon/Intron', 'PolyPhen(HDiv)', 'PolyPhen(Hvar)', 'SiftScore', 'CaddScore', 'LofScore', 'LrtScore' ]\n",
    "    denovo = pd.read_csv(readfile, sep = '\\t', header = None, names = denovo_column, comment = '#')\n",
    "    \n",
    "    \n",
    "    # filtering only 'PrimaryPhenotype' = control and find_pattern \n",
    "    denovo = denovo[denovo['Variant'].str.len() == 3]\n",
    "    denovo[np.logical_or(np.logical_and(denovo['PrimaryPhenotype'] == find_pattern, \n",
    "                                              denovo['FunctionClass'] == 'missense'),\n",
    "                               denovo['PrimaryPhenotype'] == 'control')]\n",
    "    denovo['origin'] = np.where(denovo['PrimaryPhenotype'] == find_pattern, 1,0)\n",
    "    \n",
    "    \n",
    "    # only keep exon or whole genome sequencing data\n",
    "    denovo = denovo[denovo['SequenceType'] != 'targeted']\n",
    "    denovo = denovo[np.logical_or(denovo['PrimaryPhenotype'] == find_pattern, denovo['PrimaryPhenotype'] == 'control')]\n",
    "    \n",
    "     \n",
    "    denovo['Variant'] = denovo['Variant'].str.replace('>', '-')\n",
    "    denovo = denovo.assign(location = 'chr' + denovo['Chr'].astype(str) + '-' + (denovo['Position']-1).astype(str) +\\\n",
    "                     '-' + denovo['Position'].astype(str) +'-'+ denovo['Variant'])\n",
    "    \n",
    "    \n",
    "    # remove all the unnecessary columns\n",
    "    denovo = denovo.drop(['SampleID', 'StudyName', 'PubmedID', 'NumProbands', 'NumControls',  'Chr', 'Position',\n",
    "                           'Variant', 'rsID', 'DbsnpBuild' , 'AncestralAllele', '1000GenomeCount', 'Exon/Intron',\n",
    "                     'ExacFreq', 'EspAaFreq', 'EspEaFreq', 'Transcript', 'condingDnaSize', 'Gene', 'cDnaVariant', 'ProteinVariant', \n",
    "                     'PolyPhen(HDiv)', 'PolyPhen(Hvar)', 'SiftScore', 'CaddScore', 'LofScore', 'LrtScore'], axis=1)\n",
    "    \n",
    "\n",
    "    denovo = denovo.drop_duplicates()\n",
    "    denovo_simple = denovo[['location','origin']].drop_duplicates()\n",
    "    \n",
    "    denovo_location = denovo['location'].drop_duplicates().to_frame()\n",
    "    return(denovo, denovo_simple, denovo_location)\n",
    "\n",
    "ssc = find_denovo_pattern_new('/storage/home/jkl5991/work/project/original_data/denovo-db.ssc-samples.variants.tsv.gz', 'autism')\n",
    "#ssc[2].to_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/denovo_ssc_location.tsv', header = None, sep = '\\t',index = False)\n",
    "\n",
    "nonssc = find_denovo_pattern_new('/storage/home/jkl5991/work/project/original_data/denovo-db.non-ssc-samples.variants.tsv.gz', 'developmentalDisorder')\n",
    "#nonssc[2].to_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/denovo_non_ssc_location.tsv', header = None, sep = '\\t',index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat denovo_non-ssc_location.tsv |tr '-' '\\t' > non_ssc_loc.bed \n",
    "!sortBed -i non_ssc_loc.bed > non_ssc_sortedloc.bed\n",
    "\n",
    "!cat denovo_ssc_location.tsv |tr '-' '\\t' > ssc_loc.bed \n",
    "!sortBed -i ssc_loc.bed > ssc_sortedloc.bed \n",
    "\n",
    "\n",
    "!awk '{gsub(\"chr\" , \"\"); print}' non_ssc_sortedloc.bed > non_ssc_loc4UNEECON.bed \n",
    "!awk '{gsub(\"chr\" , \"\"); print}' denovo_location_sorted.bed > denovo_location_sorted4UNEECON.bed \n",
    "\n",
    "\n",
    "# in /storage/home/jkl5991/work/project/original_data\n",
    "!tabix UNEECON_addchr.tsv.bgz -R ~/group/dbnsfp/tabix_try/ssc_sortedloc.bed > tabix_ssc_uneecon.tsv\n",
    "!tabix UNEECON_addchr.tsv.bgz -R ~/group/dbnsfp/tabix_try/non_ssc_sortedloc.bed > tabix_non_ssc_uneecon.tsv\n",
    "\n",
    "# in ~/group/dbnsfp/tabix_try\n",
    "!tabix prediction_sorted.bed.bgz -R ssc_sortedloc.bed > tabix_ssc_original.tsv \n",
    "!tabix prediction_sorted.bed.bgz -R non_ssc_sortedloc.bed > tabix_non_ssc_original.tsv \n",
    "!tabix output2.bed.bgz -R ssc_sortedloc.bed > tabix_ssc_othermodel.tsv \n",
    "!tabix output2.bed.bgz -R non_ssc_sortedloc.bed > tabix_non_ssc_othermodel.tsv \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import uneecon & original model\n",
    "# nonssc\n",
    "\n",
    "column =['chr', 'pos-1','pos', 'ref','alt','accession_num', 'SIFT_pred', 'LRT_pred', 'MA_pred', 'PROVEN_pred',\n",
    "               'SLR_score', 'SIFT_score', 'LRT_omega', 'MA_score', 'PROVEN_score', 'Grantham', 'HMMEntropy',\n",
    "               'HMMRelEntropy', 'PredRSAB', 'PredRSAI', 'PredRSAE', 'PredBFactorF', 'PredBFactorM', 'PredBFactorS',\n",
    "               'PredStabilityH', 'PredStabilityM', 'PredStabilityL', 'PredSSE', 'PredSSH', 'PredSSC', 'dscore',\n",
    "               'phyloP_pri', 'phyloP_mam', 'phyloP_ver', 'RNA_seq'] \n",
    "\n",
    "nonssc_uneecon = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_non_ssc_uneecon.tsv', sep = '\\t', names = ['chr','pos','ref','alt','UNEECON'])\n",
    "nonssc_original = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_non_ssc_original.tsv', sep = '\\t', names = column)\n",
    "nonssc_othermodel = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_non_ssc_othermodel.tsv', sep = '\\t', names = header)\n",
    "\n",
    "\n",
    "ssc_uneecon = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_ssc_uneecon.tsv', sep = '\\t', names = ['chr','pos','ref','alt','UNEECON'])\n",
    "ssc_original = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_ssc_original.tsv', sep = '\\t', names = column)\n",
    "ssc_othermodel = pd.read_csv('/storage/home/jkl5991/group/dbnsfp/tabix_try/tabix_ssc_othermodel.tsv', sep = '\\t', names = header)\n",
    "\n",
    "def denovo_process(original, uneecon, othermodel, df, semicol= ['LIST-S2_score', 'VEST4_score','MPC']):\n",
    "    \n",
    "    uneecon = createloc(uneecon)\n",
    "    original = original.drop_duplicates()\n",
    "    original = createloc(original)\n",
    "\n",
    "    uneecon['location'] = 'chr' + uneecon['location']\n",
    "    \n",
    "    \n",
    "    whole = pd.merge(original, uneecon , on = ['location'], how = 'inner')\n",
    "\n",
    "    othermodel = createloc(othermodel)\n",
    "    \n",
    "    whole = pd.merge(whole, othermodel , on = ['location'], how = 'inner')\n",
    "    \n",
    "    ## save FATHMM\n",
    "    FATHMM_score =[]\n",
    "    for i in whole['FATHMM_score'].str.split(';')[:]:\n",
    "        i = list(filter(lambda a: a != '.', i))\n",
    "        if len(i) == 0:\n",
    "            FATHMM_score.append('.')\n",
    "        else:\n",
    "            FATHMM_score.append(float(min(i))*(-1))\n",
    "    whole['FATHMM_score'] = np.array(FATHMM_score)\n",
    "    \n",
    "    whole = save_allsemicol(whole, semicol) \n",
    "\n",
    "    # remove all the rows that contain '.' (empty)\n",
    "    whole = whole.drop(columns = ['pos-1_x','pos-1_y','accession_num','integrated_fitCons_score','H1-hESC_fitCons_score'])\n",
    "    for i in features:\n",
    "        whole.drop(whole.loc[whole[i] == '.'].index, inplace= True)\n",
    "    \n",
    "    denovo = pd.merge(whole, df, on = ['location'])\n",
    "    \n",
    "    print(denovo.shape)\n",
    "    return(denovo, whole)\n",
    "\n",
    "nonssc_denovo = denovo_process(nonssc_original, nonssc_uneecon, nonssc_othermodel, nonssc[1])[0]\n",
    "ssc_denovo = denovo_process(ssc_original, ssc_uneecon, ssc_othermodel, ssc[1])[0]\n",
    "ssc_denovo_control = ssc_denovo[ssc_denovo['origin'] ==0]\n",
    "\n",
    "\n",
    "denovo = nonssc_denovo.append(ssc_denovo_control) #4229,44\n",
    "np.sum(denovo['origin'])  #3600 proband, 1300 (1183+117)control"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (sanEnv)",
   "language": "python",
   "name": "sanenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
